### PR TITLE
server: Fix compression protocol for larger packets | tidb-test=pr/2227

### LINKED ISSUE #47152: Load data with --compress=true fails
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
$ echo 'a' > /tmp/short.txt

$ mysql -h127.1 -uroot -P4000 --compress=true
WARNING: --compress is deprecated and will be removed in a future version. Use --compression-algorithms instead.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 2097154
Server version: 8.0.11-TiDB-None TiDB Server (Apache License 2.0) Community Edition, MySQL 8.0 compatible

mysql> use test
Database changed
mysql> create table t(a longtext);
Query OK, 0 rows affected (0.03 sec)

mysql> load data local infile '/tmp/short.txt' into table t;
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    2097196
Current database: test

ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    2097198
Current database: test

ERROR 2013 (HY000): Lost connection to MySQL server during query
```

### 2. What did you expect to see? (Required)

Success

### 3. What did you see instead (Required)

If the file exists: the client loses connection and TiDB logs `invalid compressed sequence 2 != 1`
If the file doesn't exist: the client blocks and TiDB logs `not drained yet, try reading left data from client connection`.

### 4. What is your TiDB version? (Required)

tidb: master

```
$ mysql --version
mysql  Ver 8.1.0 for macos12.6 on x86_64 (Homebrew)
```

### LINKED ISSUE #47157: Send huge SQL with --compress=true fails
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
# create a huge SQL
$ mysql -h127.1 -uroot -P4000
mysql> select concat('select ', repeat('a', 16*1024*1024)) into outfile '/tmp/data.txt';

$ mysql -h127.1 -uroot -P4000 --compress=true < /tmp/data.txt;
WARNING: --compress is deprecated and will be removed in a future version. Use --compression-algorithms instead.
ERROR 2013 (HY000) at line 1: Lost connection to MySQL server during query
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

Lost connection and TiDB logs `unexpected EOF`.

### 4. What is your TiDB version? (Required)

tidb: master

```
$ mysql --version
mysql  Ver 8.1.0 for macos12.6 on x86_64 (Homebrew)
```

### LINKED ISSUE #47161: Querying huge data with useCompress=true fails
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```java
        String url = "jdbc:mysql://127.1:4000/test?useCompression=true";
        String user = "root";

        Class.forName("com.mysql.cj.jdbc.Driver");

        Connection conn = null;
        ResultSet rs = null;
        try {
            conn = DriverManager.getConnection(url, user, "");
            Statement stmt = conn.createStatement();
            rs = stmt.executeQuery("select repeat('a', 16*1024*1024)");
            rs.next();
            System.out.println(rs.getString(1).length());
            stmt.close();
            rs.close();
        } finally {
            if (conn != null) conn.close();
        }
```

### 2. What did you expect to see? (Required)

Output 16777216

### 3. What did you see instead (Required)

```
Exception in thread "main" java.sql.SQLException: Packets received out of order
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeQuery(StatementImpl.java:1200)
	at Charset.main(Charset.java:15)
```

### 4. What is your TiDB version? (Required)

tidb: master
JDBC: 8.01

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number:  close #47152

All issues fixed by this PR:
* close #47152
* close #47157
* close #47161

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
A bug in the handling of the compressed protocol for larger payloads (>=16M) has been fixed
```

### CODE REVIEW

**Comment by dveeden** [comment]
```perl
#!/bin/perl
use strict;
use DBI;
use v5.36.0;

$| = 1;

my @dsns = (
	'DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5',
	'DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5;mysql_compression=1'
);

foreach my $dsn (@dsns) {
	say "Testing with DSN: " . $dsn;
	my $dbh = DBI->connect($dsn, 'root', undef);

	my $sth = $dbh->prepare("select repeat('a', ?)");
	my $size = (2**24)-20;
	while ($size < (2**24)+10) {
		print "Trying " . $size;
		$sth->execute($size) or die "FAIL: query execution failed\n";
		while (my $ref = $sth->fetchrow_arrayref()) {
			if (length($$ref[0]) == $size) {
				print " OK\n";
			} else {
				print " FAIL: expected " . $size . " but got " . $$ref[0] . "\n";
			}
			$size++;
		}
	}
	$sth->finish;
}
```

With this PR:
```
$ perl ~/compress_large_issue47161.pl 
Testing with DSN: DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5
Trying 16777196 OK
Trying 16777197 OK
Trying 16777198 OK
Trying 16777199 OK
Trying 16777200 OK
Trying 16777201 OK
Trying 16777202 OK
Trying 16777203 OK
Trying 16777204 OK
Trying 16777205 OK
Trying 16777206 OK
Trying 16777207 OK
Trying 16777208 OK
Trying 16777209 OK
Trying 16777210 OK
Trying 16777211 OK
Trying 16777212 OK
Trying 16777213 OK
Trying 16777214 OK
Trying 16777215 OK
Trying 16777216 OK
Trying 16777217 OK
Trying 16777218 OK
Trying 16777219 OK
Trying 16777220 OK
Trying 16777221 OK
Trying 16777222 OK
Trying 16777223 OK
Trying 16777224 OK
Trying 16777225 OK
Testing with DSN: DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5;mysql_compression=1
Trying 16777196 OK
Trying 16777197 OK
Trying 16777198 OK
Trying 16777199 OK
Trying 16777200 OK
Trying 16777201 OK
Trying 16777202 OK
Trying 16777203 OK
Trying 16777204 OK
Trying 16777205 OK
Trying 16777206 OK
Trying 16777207 OK
Trying 16777208 OK
Trying 16777209 OK
Trying 16777210 OK
Trying 16777211 OK
Trying 16777212 OK
Trying 16777213 OK
Trying 16777214 OK
Trying 16777215 OK
Trying 16777216 OK
Trying 16777217 OK
Trying 16777218 OK
Trying 16777219 OK
Trying 16777220 OK
Trying 16777221 OK
Trying 16777222 OK
Trying 16777223 OK
Trying 16777224 OK
Trying 16777225 OK
```

Without this PR:
```
$ perl ~/compress_large_issue47161.pl 
Testing with DSN: DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5
Trying 16777196 OK
Trying 16777197 OK
Trying 16777198 OK
Trying 16777199 OK
Trying 16777200 OK
Trying 16777201 OK
Trying 16777202 OK
Trying 16777203 OK
Trying 16777204 OK
Trying 16777205 OK
Trying 16777206 OK
Trying 16777207 OK
Trying 16777208 OK
Trying 16777209 OK
Trying 16777210 OK
Trying 16777211 OK
Trying 16777212 OK
Trying 16777213 OK
Trying 16777214 OK
Trying 16777215 OK
Trying 16777216 OK
Trying 16777217 OK
Trying 16777218 OK
Trying 16777219 OK
Trying 16777220 OK
Trying 16777221 OK
Trying 16777222 OK
Trying 16777223 OK
Trying 16777224 OK
Trying 16777225 OK
Testing with DSN: DBI:mysql:database=test;port=4000;host=127.0.0.1;mysql_read_timeout=5;mysql_compression=1
Trying 16777196 OK
Trying 16777197 OK
Trying 16777198 OK
Trying 16777199 OK
Trying 16777200 OK
Trying 16777201 OK
Trying 16777202 OK
Trying 16777203 OK
Trying 16777204 OK
Trying 16777205 OK
Trying 16777206 OK
Trying 16777207DBD::mysql::st execute failed: Lost connection to MySQL server during query at /home/dvaneeden/compress_large_issue47161.pl line 21.
FAIL: query execution failed
```

**Review by djshow832 — COMMENTED** [review]
1. Can you add a unit test to prevent it from broken by others?
2. As mentioned in https://github.com/pingcap/tidb/issues/47161#issuecomment-1742322274, what if the packet size is larger than 256MB?

**Comment by dveeden** [comment]
> 1. Can you add a unit test to prevent it from broken by others?

Yes. That was why the PR was still in draft status. 

>     2. As mentioned in [Querying huge data with useCompress=true fails #47161 (comment)](https://github.com/pingcap/tidb/issues/47161#issuecomment-1742322274), what if the result set is larger than 256MB?

That still needs to be fixed.

**Comment by djshow832** [comment]
> That still needs to be fixed.
In another PR or in this PR later?

**Comment by dveeden** [comment]
> > That still needs to be fixed.
> 
> In another PR or in this PR later?

After looking at this again I don't think this needs additional work. This is because the compressed packet id wraps around. So sending a large amount of data in small packets should be fine.

File: server/internal/packetio_test.go [hunk 1]
```diff
@@ -62,6 +62,31 @@ func TestPacketIOWrite(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = pkt.Flush()
+	require.NoError(t, err)
+
+	compressedLength := []byte{0x18, 0x4, 0x0} // 1048 bytes
+	packetNr := []byte{0x0}
+	uncompressedLength := []byte{0x0, 0x0, 0x10} // 1048576 bytes
+
+	require.Equal(t, compressedLength, testdata.Bytes()[:3])
```
Comment (djshow832) [inline server/internal/packetio_test.go:85]: How do you know that the length of the uncompressed packet is 16M?

File: pkg/server/internal/packetio.go [hunk 2]
```diff
@@ -155,13 +155,15 @@ func (p *PacketIO) readOnePacket() ([]byte, error) {
...
 		p.compressedWriter.compressedSequence = p.compressedSequence
+		compressedLength := int(uint32(compressedHeader[0]) | uint32(compressedHeader[1])<<8 | uint32(compressedHeader[2])<<16)
 		uncompressedLength := int(uint32(compressedHeader[4]) | uint32(compressedHeader[5])<<8 | uint32(compressedHeader[6])<<16)
 
 		if uncompressedLength > 0 {
 			switch p.compressionAlgorithm {
 			case mysql.CompressionZlib:
 				var err error
-				r, err = zlib.NewReader(p.bufReadConn)
+				lr := io.LimitReader(p.bufReadConn, int64(compressedLength))
```
Comment (djshow832) [inline pkg/server/internal/packetio.go:165]: What about zstd?

Comment (dveeden) [reply pkg/server/internal/packetio.go:165]: Unfortunately we can't use the same test for ZSTD as MySQL Connector/J doesn't support this: https://bugs.mysql.com/bug.php?id=112732

I created a similar test for this in C:
```c
#include <mysql.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define COMPRESS_NONE 0
#define COMPRESS_ZLIB 1
#define COMPRESS_ZSTD 2

void test_compression(int compression, size_t payload_size) {
  static MYSQL *conn;
  MYSQL_STMT *stmt;
  MYSQL_BIND bind[1];
  MYSQL_RES *res;
  MYSQL_ROW row;
  long unsigned int length;

  conn = mysql_init(NULL);
  if (conn == NULL) {
    fprintf(stderr, "mysql_init() failed\n");
    exit(1);
  }

  switch (compression) {
  case COMPRESS_NONE:
    printf("testing with no compression and a payload of %zu.\n", payload_size);
    break;
  case COMPRESS_ZLIB:
    printf("testing with ZLIB and a payload of %zu.\n", payload_size);
    mysql_options(conn, MYSQL_OPT_COMPRESSION_ALGORITHMS, "zlib");
    break;
  case COMPRESS_ZSTD:
    printf("testing with ZSTD and a payload of %zu.\n", payload_size);
    mysql_options(conn, MYSQL_OPT_COMPRESSION_ALGORITHMS, "zstd");
    break;
  default:
    fprintf(stderr, "Error: no valid compression method specified.\n");
    exit(1);
  }
  uint ssl_mode = SSL_MODE_DISABLED;
  mysql_options(conn, MYSQL_OPT_SSL_MODE, &ssl_mode);

  if (mysql_real_connect(conn, "127.0.0.1", "root", "", "test", 4000, NULL,
                         0) == NULL) {
    fprintf(stderr, "Failed to connect to the database.\n");
    if (conn != NULL)
      fprintf(stderr, "Error: %s\n", mysql_error(conn));
    mysql_close(conn);
    exit(1);
  }

  const char *sql_to_send = "INSERT INTO BLOBTEST(blobdata) VALUES (?)";
  size_t required_size =
      payload_size - strlen(sql_to_send) - strlen("_binary ''");

  char buf[required_size];
  for (size_t i = 0; i < required_size; i++) {
    buf[i] = 65 + rand() % 26;
  }
  memset(bind, 0, sizeof(bind));
  bind[0].buffer_type = MYSQL_TYPE_BLOB;
  bind[0].length = &length;
  bind[0].is_null = 0;

  mysql_query(conn, "SET GLOBAL max_allowed_packet=6291456"); //  1024 * 1024 * 6
  mysql_query(conn, "DROP TABLE IF EXISTS BLOBTEST");
  mysql_query(conn, "CREATE TABLE BLOBTEST (pos int PRIMARY KEY "
                    "auto_increment, blobdata LONGBLOB)");

  stmt = mysql_stmt_init(conn);
  if (mysql_stmt_prepare(stmt, sql_to_send, strlen(sql_to_send))) {
    printf("Failed to prepare statement: %s\n", mysql_stmt_error(stmt));
    return;
  }
  if (mysql_stmt_bind_param(stmt, bind)) {
    printf("Failed to bind argument for prepared statement: %s\n",
           mysql_stmt_error(stmt));
    return;
  };
  size_t pos = 0;
  while (pos < strlen(buf)) {
    char sbuf[4096];
    size_t slen = sizeof(sbuf);
    if ((pos + slen) > strlen(buf)) {
	    slen = strlen(buf) - pos;
    }
    // printf("pos=%zu, strlen(buf)=%zu, slen=%zu\n", pos, strlen(buf), slen);
    memcpy(sbuf, buf + pos, slen);
    if (mysql_stmt_send_long_data(stmt, 0, sbuf, sizeof(sbuf))) {
      printf("Failed to send long data for prepared statement: %s\n",
             mysql_stmt_error(stmt));
      return;
    };
    pos += sizeof(sbuf);
  }

  if (mysql_stmt_execute(stmt)) {
    printf("Failed to execute prepared statement: %s\n",
           mysql_stmt_error(stmt));
    return;
  };

  if (mysql_query(conn, "SELECT blobdata from BLOBTEST LIMIT 1")) {
    printf("Failed to query blobdata: %s\n", mysql_error(conn));
    return;
  }

  res = mysql_store_result(conn);
  row = mysql_fetch_row(res);
  if (row != NULL) {
    if (strncmp(row[0], buf, strlen(buf)) != 0) {
      fprintf(stderr, "Buffer and result don't match.\n");
      return;
    }
  }

  mysql_close(conn);
}

int main() {
  srand(time(NULL));
  printf("Testing compression\n");

  test_compression(COMPRESS_NONE, 1024 * 1024 * 3 - 2);
  test_compression(COMPRESS_NONE, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_NONE, 1024 * 1024 * 6);

  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 2 - 3);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 2 - 2);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 6);

  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 2 - 3);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 2 - 2);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 6);

  mysql_library_end();
  exit(0);
}
```

And running this against master (so without this PR) returns this:
```
$ gcc $(mysql_config --cflags --libs) -Wall -Wextra mysql-compression-test.c -std=gnu18 -g -o mysql-compression-test && ./mysql-compression-test 
Testing compression
testing with no compression and a payload of 3145726.
testing with no compression and a payload of 3145727.
testing with no compression and a payload of 6291456.
testing with ZLIB and a payload of 3145723.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 3145724.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 3145727.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 6291456.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZSTD and a payload of 3145723.
testing with ZSTD and a payload of 3145724.
testing with ZSTD and a payload of 3145727.
testing with ZSTD and a payload of 6291456.
```

This is with this config:
```toml
[performance]
txn-entry-size-limit=125829120
```
I don't think the Java test required this, so the test might not be identical.

With this PR the output is like this:
```
$ gcc $(mysql_config --cflags --libs) -Wall -Wextra mysql-compression-test.c -std=gnu18 -g -o mysql-compression-test && ./mysql-compression-test 
Testing compression
testing with no compression and a payload of 3145726.
testing with no compression and a payload of 3145727.
testing with no compression and a payload of 6291456.
testing with ZLIB and a payload of 3145723.
testing with ZLIB and a payload of 3145724.
testing with ZLIB and a payload of 3145727.
testing with ZLIB and a payload of 6291456.
testing with ZSTD and a payload of 3145723.
testing with ZSTD and a payload of 3145724.
testing with ZSTD and a payload of 3145727.
testing with ZSTD and a payload of 6291456.
```

So looks like this doesn't affect ZSTD. 

Do you think I should still apply the same change for ZSTD, @djshow832 ?

Comment (djshow832) [reply pkg/server/internal/packetio.go:165]: My intention is to send a packet longer than 16MB.

**Review by lance6716 — APPROVED** [review]

**Review by djshow832 — COMMENTED** [review]
I don't see how the second problem in https://github.com/pingcap/tidb/issues/47152#issuecomment-1729106547 is solved: sometimes uncompressed sequence is affected by compressed sequence.

**Comment by dveeden** [comment]
> I don't see how the second problem in [#47152 (comment)](https://github.com/pingcap/tidb/issues/47152#issuecomment-1729106547) is solved: sometimes uncompressed sequence is affected by compressed sequence.

With MySQL 8.2.0:
image

| Packet | Sequence | Compressed Sequence | Description |
|---|---|---|---|
|24 | 0 | 0 | Query: LOAD DATA INFILE |
|25 | 1 | 1 | LOCAL INFILE request |
| 27 | 2,3 | 2 | LOCAL INFILE response |
|28 | 3 | 3 | OK |

With TiDB (this PR):
image
| Packet | Sequence | Compressed Sequence | Description |
|---|---|---|---|
|57 | 0 | 0 | Query: LOAD DATA INFILE |
|58 | 1 | 1 | LOCAL INFILE request |
| 60 | 2,3 | 2 | LOCAL INFILE response |
|61 | 4 | 3 | OK |

Here the sequence number 3 gets duplicated with MySQL, probably because in `net_flush()` for packet 27 the compressed packet number (`2`) gets synced to the regular packet number. 

I'm not sure which one is best, but I also haven't seen any issues with this from a client standpoint. I had a quick look at the Connector/J code and there I didn't see packet numbers of the regular and compressed packets being synced.

**Comment by dveeden** [comment]
> I don't see how the second problem in [#47152 (comment)](https://github.com/pingcap/tidb/issues/47152#issuecomment-1729106547) is solved: sometimes uncompressed sequence is affected by compressed sequence.

I've now done this in ee667790b613c7f643422ffcb27fbaa52c036c45

**Review by djshow832 — APPROVED** [review]

**Pull request merged**